### PR TITLE
Copy Site: Add back functionality on domains step

### PR DESCRIPTION
**Pull request by kozer**
See p4TIVU-aoH-p2#comment-11701
Related to https://github.com/Automattic/dotcom-forge/issues/1519

Original pr: https://github.com/Automattic/wp-calypso/pull/72327

#### Proposed Changes

Refactor of the `DomainsStep` component to solve the bug where the user needs to click twice to the bug button to go back to the `/sites` page.

The step has been split into two components: the main component which renders the step wrapper (which includes the header), and a child component which renders the domain picker. The component that renders the header no longer needs to subscribe to the `domainForm` value from the onboarding store. Each time the domain search field is unfocused `domainForm` gets resaved, and this is what was causing the header to rerender.

##### Test Copy site flow
1. Make sure you have a site with `Business` plan and atomic.
2. Navigate to `/sites` 
3. Click the dots menu on the right (at your site's row)
4. You should be navigated to the domain selection screen.
5. On the top right there is a `Back` button. Click it **once**, and ensure that you navigate back to the `/sites` page

##### Test Link in bio flow

Because of the refactor, makes sense to take a look at `link-in-bio` flow as well to see that the component works as expected:

1. Navigate to `/setup?flow=link-in-bio`
2. Take the flow until you land to the domain selection screen.
3. Ensure that the input is working (search for a domain and ensure that it works)
4. Test that the `Decide later` button works (should skip this step)
5. Make sure that the filters are working.
6. Select a free domain and ensure that the process completes successfully.
7. Ensure you select a free domain and a plan other than Free. Make sure you can checkout successfully.

#### Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

fixes https://github.com/Automattic/dotcom-forge/issues/1519

### CODE REVIEW

**Comment by ciampo** [comment]
Hey @kozer , us (Team Components) usually work on [the `@wordpress/components` package in Gutenberg](https://github.com/WordPress/gutenberg/blob/trunk/packages/components) and therefore don't have a lot of familiarity with Calypso and its onboarding/landing areas. Is there anyone else that you could tag to help review this PR?

Thank you!

**Comment by markbiek** [comment]
## Testing Copy Site flow

When I clicked the dots next too my site, I got a drop down menu instead of being taken to the domain selection screen.

image

I thought maybe I was supposed to click "Copy" so I did and _that_ took me to the domain selection screen.

Unfortunately there was no back button:
image

## Testing link-in-bio flow

This worked as expected.

**Comment by p-jackson** [comment]
@kozer The original refactor was a bit messy so I've tidied up a bit by moving all the domain-picking code into the child component. I've also split each component into its own file.

I've also removed the code which tidied up the `onMouseDown` hack. I think that should be done in #72327 rather than this PR. We could actually rebase this PR now so that it's based on `trunk` if you wanted. Up to you.

@Automattic/vertex FYI we're refactoring the domain step to remove problematic re-renders. There shouldn't be any changes in behaviour, just want to give you a heads-up and an opportunity to take a look at the change.

**Comment by kozer** [comment]
@p-jackson 

> @kozer The original refactor was a bit messy so I've tidied up a bit by moving all the domain-picking code into the child component. I've also split each component into its own file.
> 
> I've also removed the code which tidied up the `onMouseDown` hack. I think that should be done in #72327 rather than this PR. We could actually rebase this PR now so that it's based on `trunk` if you wanted. Up to you.
> 
> @Automattic/vertex FYI we're refactoring the domain step to remove problematic re-renders. There shouldn't be any changes in behaviour, just want to give you a heads-up and an opportunity to take a look at the change.

@p-jackson as this branch contains all the changes I did on my branch and solve the issue, I don't see the value of having my branch open. I'll rebase this pr, update the description, and close mine.

I'll also remove the `onMouseDown` and have this as the main branch that solves the issue.

Thank you very much for your contribution.

**Review by p-jackson — APPROVED** [review]
Caveat: I did the refactor so take this review with a grain of salt.

I've tested the domain step and I can't find any regressions so I think the refactor is good 🤞

The other flows I found that use this step are `/setup/ecommerce` and `/setup/link-in-bio-tld` and I couldn't find a way to break them either.

I'd hoped to test the case where you already own the domain you want to use, but afaict none of these flows supported domain mapping to begin with.

**Comment by kozer** [comment]
> I've tested the domain step and I can't find any regressions so I think the refactor is good crossed_fingers

Me neither. I guess we are good. I hope that on Monday others will check it as well, so we will be more confident.

File: client/landing/stepper/declarative-flow/internals/steps-repository/domains/index.tsx [hunk 1]
```diff
@@ -291,186 +204,30 @@ const DomainsStep: Step = function DomainsStep( { navigation, flow } ) {
...
+		/>
+	);
 
 	return (
 		<StepContainer
 			stepName="domains"
 			isWideLayout={ true }
-			hideBack={ true }
+			shouldStickyNavButtons={ isCopySiteFlow( flow ) }
+			hideBack={ ! isCopySiteFlow( flow ) }
```
Comment (sejas) [inline client/landing/stepper/declarative-flow/internals/steps-repository/domains/index.tsx:225]: We can consider changing the text to "Back to Sites".

```suggestion
			hideBack={ ! isCopySiteFlow( flow ) }
			backLabelText={ __( 'Back to Sites' ) }
```
Screenshot 2023-02-06 at 13 20 08

**Review by sejas — APPROVED** [review]
I tested it and it works as expected. I couldn't find any regression due the big refactor. I tried `/setup/link-in-bio` nor `/setup/ecommerce` and those still working as before.

My only comment is that this back button is different from import flow:
https://github.com/Automattic/wp-calypso/pull/72327#issuecomment-1401555490

We could consider changing the text to `Back to Sites`. It will improve the experience when the flow is triggered from `Settings General > Site Tools`.
Screenshot 2023-02-06 at 13 26 56

**Comment by wojtekn** [comment]
I've rechecked it, and I see two things:

- we use different label and text for "Back" button in different flows

| Standard flow | Copy Site flow |
|----|----|
| Screenshot 2023-02-06 at 14 26 52 | Screenshot 2023-02-06 at 14 27 41 |

- when I load/reload domain step in Copy Site flow, "Copy Site" appears near WP logo for a short time, and then disappears

Screen Capture on 2023-02-06 at 14-33-09

**Comment by kozer** [comment]
> we use different label and text for "Back" button in different flows

Fixed.

>when I load/reload domain step in Copy Site flow, "Copy Site" appears near WP logo for a short time, and then disappears

This is how the `StepContainer` works. Let's keep the scope limited, and open another issue for that.

**Comment by danielbachhuber** [comment]
> This is how the `StepContainer` works. Let's keep the scope limited, and open another issue for that.

@kozer Can you open this issue?

**Comment by danielbachhuber** [comment]
@kozer Also, given this is a somewhat risky refactor, what are the key metrics we should monitor post-deploy?

**Comment by kozer** [comment]
> @kozer Also, given this is a somewhat risky refactor, what are the key metrics we should monitor post-deploy?

Taking a look at the code, I found the following events,  that we can check to see if there is a huge discrepancy between the pre and post-deploy.

It's the `ga` event with `category`: `Domain Search`
and the following events on our side: `calypso_domain_search_submit_step`, `'calypso_signup_skip_step'`

**Comment by danielbachhuber** [comment]
> Taking a look at the code, I found the following events, that we can check to see if there is a huge discrepancy between the pre and post-deploy.

@kozer Great, thank you. Can you report the results to `#dotcom-yolo` in a few hours, and again tomorrow?

When you do, you can link to the Slack messages from here. It's better to report the data in Slack so we can keep it private.

**Comment by kozer** [comment]
@danielbachhuber , today I'm leaving a bit early unfortunately. Sorry about that. Tomorrow I will be on top of that all day

**Comment by danielbachhuber** [comment]
@kozer Ok. In the future, please make sure you're around for at least a few hours after landing a major code change.

**Comment by danielbachhuber** [comment]
@kozer Another thing I just noticed (that's apparent in @wojtekn's screenshot too): the action bar has a white background, even though the domains background is light gray

image

**Comment by danielbachhuber** [comment]
> the action bar has a white background, even though the domains background is light gray

Created a follow-up issue for this: https://github.com/Automattic/dotcom-forge/issues/1622

**Comment by kozer** [comment]
> @kozer Another thing I just noticed (that's apparent in @wojtekn's screenshot too): the action bar has a white background, even though the domains background is light gray
> 
> image
> 

@danielbachhuber I agree, but this is not relevant with the scope of the pr. This is how it was implemented originally. Do you want me to create a new issue for that?

**Pull request merged**